pkg/controller/drain/drain_controller.go

```diff
@@ -347,6 +354,32 @@ func (ctrl *Controller) syncNode(key string) error {
 		return fmt.Errorf("node %s: non-recognized drain verb detected %s", node.Name, desiredVerb)
 	}
 
+	// TEST: finally, let's drain ourself so we don't have to wait for the restart
```

Comment:
Isn't this just a very roundabout way to:

- Release the lease
- `exit(0)`

Hmm except if we do this aren't we losing out on the completion annotation below?

Reply:
More concretely, the behaviour I've observed without this PR:

1. the master node this controller is running on is marked for update
2. the daemon tells the controller to drain, the controller does, the daemon then completes the update and reboots the node
3. the controller pod is listed as "running" still, but it is supposed to be running on that master node that just shut down. The pod doesn't get scheduled anywhere else

With this PR:
1 and 2 are the same, but instead of completing the drain and then having the daemon do its job, the controller will drain itself to another node, and start immediately. The new controller picks up, marks the completion, and the daemon on the old node continues while the new controller chugs on and does other activities (such as drain workers)

Reply:
> if we exit instead of drain, the controller pod will simply stay dead until the master node it was supposed to schedule on reboots.

Won't kube reschedule the pod when it exits?  And it should happen on another node because the current one isn't schedulable.

Although officially draining would avoid a potential race where if we just exit, but then reboot, kubelet may go down before the pod status is updated.

Hmm.  I think the graceful node shutdown is really the fix for these races.

Reply:
With John's changes in https://github.com/openshift/machine-config-operator/pull/3185/files#diff-98a2adaa91130abb8fb7321e0ed6de823657c91b48c10e67c141a12d0a0197feR108 we should already be exiting right? The pod never gets rescheduled. I don't know what the expected behaviour is though, I should check that

Reply:
Side note: I think this may also work with context-ifying all the controllers and properly handling their shutdowns so we get rescheduled, but directly calling the drain is "easier" in a sense here

Reply:
I realized that this is the equivalent of removing the filter to not drain the MCC pod, so let's see if this works